# #ARTIFICIAL SELECTION
# 
# # #
# # sim10n_mono_1of2.py
# # 
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# # 
# # #
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM  	  
# # #
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = 500)
# 
# 		
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_043414
# 
# ##
# sim10n_mono.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	  
# #
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# # Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_043414
# 
# # #
# # sim10n_mono.py
# # 
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# # 
# # #
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# # #
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# # Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_043414
# 
# ##NATURAL SELECTION
# 
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_a_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_b_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_c_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_d_043414
  
##
# sim10n_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_e_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_f_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_g_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_h_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_i_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_j_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_a_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_b_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_c_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_d_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_e_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_f_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_g_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_h_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_i_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_j_043414

#ARTIFICIAL SELECTION

# #
# sim10n_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM  	  
# #

# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_a_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_a_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_a_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = 500)
# 
# 		
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_b_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_b_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_b_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_c_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_c_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_c_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_d_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_d_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_d_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_e_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_e_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_e_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_f_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_f_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_f_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_g_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_g_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_g_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_h_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_h_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_h_043414
# 
# # #
# # sim10n_mono.py
# # 
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# # 
# # #
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# # #
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# # Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_i_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_i_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_i_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_000000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_000000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_040000_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_040000_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_001400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_001400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_000034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_000034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_041400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_041400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_041434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_041434_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_040034_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_040034_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_140400_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_140400_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_340004_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_340004_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_j_043414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_j_043414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_j_043414
# 
##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_a_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_a_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_a_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_b_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_b_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_b_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_c_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_c_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_c_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_d_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_d_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_d_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_e_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_e_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_e_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_f_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_f_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_f_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_g_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_g_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_g_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_h_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_h_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_h_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_i_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_i_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_i_043414
# 
# ##
# # sim10n_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# # 		sim.DiscardIf(True, subPops=[
# # 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_j_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_j_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_j_043414

# ##NATURAL SELECTION
# 
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_a_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_a_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_a_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_a_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_b_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_b_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_b_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_b_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_c_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_c_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_c_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_c_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_d_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_d_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_d_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_d_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_e_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_e_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_e_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_e_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_f_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_f_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_f_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_f_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_g_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_g_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_g_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_g_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_h_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_h_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_h_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_h_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_i_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_i_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_i_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_i_043414
#   
# ##
# # sim10n_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_1of2_nat_j_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_1of2_nat_j_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_1of2_nat_j_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_1of2_nat_j_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_a_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_a_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_a_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_a_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_b_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_b_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_b_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_b_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_c_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_c_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_c_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_c_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_d_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_d_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_d_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_d_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_e_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_e_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_e_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_e_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_f_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_f_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_f_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_f_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_g_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_g_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_g_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_g_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_h_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_h_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_h_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_h_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_i_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_i_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_i_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_i_043414
#   
# ##
# # sim10n_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.0005 # Forward Mutation Rate
# v = 0.00005# Backward Mutation Rate
# m = 0.00005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_000000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_040000
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_001400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# # 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_000034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_041400
# 
# ###################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_041434
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_040034
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 300   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_140400
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 80    # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# #       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_340004
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 80   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10n_mono_nat_j_341404
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 300   # Number of Steps (Generations) before selection event 2
# j3 = 200   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10n_mono_nat_j_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10n_mono_nat_j_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# #sim10n_mono_nat_j_043414

# #ARTIFICIAL SELECTION
# 
# # #
# # sim10p_mono_1of2.py
# # 
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# # 
# # #
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: April 11, 2013 01:41:29 PM   	  
# # #
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10p_mono_1of2_himu_a_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10p_mono_1of2_himu_a_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10p_mono_1of2_himu_a_340414
# 
# ##
# # sim10p_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: April 11, 2013 01:41:29 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10p_mono_himu_a_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10p_mono_himu_a_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10p_mono_himu_a_341404
# 
# ##NATURAL SELECTION
# 
# ##
# # sim10p_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10p_mono_1of2_nat_himu_a_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10p_mono_1of2_nat_himu_a_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# 		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# 		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10p_mono_1of2_nat_himu_a_340414
# 
# ##
# # sim10p_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: April 11, 2013 01:50:21 PM 	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10p_mono_nat_himu_a_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10p_mono_nat_himu_a_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# 	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# 	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10p_mono_nat_himu_a_340414
# 
# #ARTIFICIAL SELECTION
# 
# # #
# # sim10p_mono_1of2.py
# # 
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# # 
# # #
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: April 11, 2013 01:41:29 PM   	  
# # #
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10p_mono_1of2_himu_b_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10p_mono_1of2_himu_b_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10p_mono_1of2_himu_b_340414
# 
# ##
# # sim10p_mono.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: April 11, 2013 01:41:29 PM	  
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10p_mono_himu_b_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10p_mono_himu_b_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10p_mono_himu_b_341404
# 
# ##NATURAL SELECTION
# 
# ##
# # sim10p_mono_1of2_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: February 25, 2013 12:24:31 PM	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10p_mono_1of2_nat_himu_b_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10p_mono_1of2_nat_himu_b_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# 		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# 		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10p_mono_1of2_nat_himu_b_340414
# 
# ##
# # sim10p_mono_nat.py
# #
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # # 1 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# 
# # genepop, fstat format
# # 
# #
# ##
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: April 11, 2013 01:50:21 PM 	   
# ##
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 1   # Initial Number of Subpopns
# d = 6 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# 
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10p_mono_nat_himu_b_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10p_mono_nat_himu_b_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
# 	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
# 	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10p_mono_nat_himu_b_340414
# 
# #ARTIFICIAL SELECTION
# 
# # #
# # sim10p_mono_1of2.py
# # 
# # Script to simulate a population with n subpops of i individuals, with population rebound, 
# # natural selection and sampling.  Selection occuring at different periods.
# # 2 Initial subpopns, expanding to 6
# # output as genotypes for Powermarker
# # Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# # If other selection events before that at locus 4, this occurs before the split also
# # 
# # genepop, fstat format
# # 
# # #
# # Author: Richard Stephens
# # Created: October 29, 2012  08:56:33 AM
# # Modified: April 11, 2013 01:41:29 PM   	  
# # #
# 
# 
# import simuOpt
# import simuPOP as sim
# import math
# from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
# from simuPOP.sampling import drawRandomSample
# 
#        
# ## Settings ###################################################################################
# n = 2   # Initial Number of Subpopns
# d = 3 # Divisor for subpopn splitting
# i = 100000 # Number of Indivs/Subpopn
# l = 10    # Number of Loci per Chromosome
# c = 5 # Number of Chromosomes
# g = 10 # Number of Steps (Generations) before expansion
# t = 2000 # Total Number of Steps (Generations)
# u = 0.005 # Forward Mutation Rate
# v = 0.0005# Backward Mutation Rate
# m = 0.0005  # Migration rate
# e = 98 # Proportion of selfing
# s1 = 0.008 # selection coefficient
# q = 100000 # Maximum Population Size
# r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)
# 
# ##################################################################################################
# 
# z1 = 120   # Number of Steps (Generations) before population splitting
# j1 = 100   # Number of Steps (Generations) before selection event 1
# j2 = 200   # Number of Steps (Generations) before selection event 2
# j3 = 60   # Number of Steps (Generations) before selection event 3
# 
# d1 = 50 # duration of selection event 1 (in generations)
# d2 = 50 # duration of selection event 2 (in generations)
# d3 = 50 # duration of selection event 3 (in generations)
# 
# ##################################################################################################
# 
# pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
# pop.setVirtualSplitter(sim.ProductSplitter([
#     sim.AffectionSplitter()
#     ])
# )
# z=20
# def demo(pop,gen):
# 	global q, g
# 	rate=2**15
# 	# if .... stop growing
# 	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
# 		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
# 	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
# 		return ([q for x in pop.subPopSizes(ancGen=-1)])
# 	else:
# 		return ([x for x in pop.subPopSizes(ancGen=-1)])
# 		
# def sampleAndExport(pop):
#     sz = pop.subPopSizes()
#     new_sz = [x//200 for x in sz]
#     sample = drawRandomSample(pop, new_sz)
#     export(sample, format='genepop', output='sim10p_mono_1of2_himu_c_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
#     export(sample, format='fstat', output='sim10p_mono_1of2_himu_c_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
#     return True
# 
# simu = sim.Simulator(pop, rep=1)
# simu.evolve(
#     initOps=[
# 	    sim.InitGenotype(freq=(0.5, 0.5)),
# 	],
# 	preOps=[
# 		sim.SNPMutator(u = u, v = v),
# 		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
# 		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
# 	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#         sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#         sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
#     ],
# 	
#     #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
#     matingScheme=sim.HeteroMating([
#         sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
#         sim.HomoMating(
#             chooser=sim.CombinedParentsChooser(
#                 sim.RandomParentChooser(),
#                 sim.RandomParentChooser()),
#             generator=sim.OffspringGenerator(
#                 ops= [
# 	                sim.MendelianGenoTransmitter(),
# 	                sim.Recombinator(intensity=r1)
#                 ],
#             ),
#             weight=100-e)
#         ],
#     subPopSize=demo
#     ),
# 	postOps=[
# 		sim.PyOperator(func=sampleAndExport, at = [500,t])
#     ],
# 	gen=(t+1),
# )
# print 'all done'
# ##sim10p_mono_1of2_himu_c_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_himu_c_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_himu_c_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_himu_c_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_himu_c_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_himu_c_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_himu_c_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_himu_c_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_himu_c_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_himu_c_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_himu_d_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_himu_d_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_himu_d_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_himu_d_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_himu_d_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_himu_d_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_himu_d_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_himu_d_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_himu_d_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_himu_d_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_himu_d_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_himu_d_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_himu_e_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_himu_e_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_himu_e_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_himu_e_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_himu_e_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_himu_e_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_himu_e_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_himu_e_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_himu_e_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_himu_e_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_himu_e_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_himu_e_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_himu_f_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_himu_f_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_himu_f_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_himu_f_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_himu_f_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_himu_f_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_himu_f_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_himu_f_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_himu_f_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_himu_f_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_himu_f_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_himu_f_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_himu_g_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_himu_g_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_himu_g_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_himu_g_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_himu_g_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_himu_g_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_himu_g_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_himu_g_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_himu_g_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_himu_g_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_himu_g_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_himu_g_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_himu_h_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_himu_h_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_himu_h_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_himu_h_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_himu_h_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_himu_h_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_himu_h_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_himu_h_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_himu_h_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_himu_h_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_himu_h_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_himu_h_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_himu_i_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_himu_i_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_himu_i_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_himu_i_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_himu_i_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_himu_i_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_himu_i_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_himu_i_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_himu_i_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_himu_i_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_himu_i_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_himu_i_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_himu_j_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_himu_j_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_himu_j_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_himu_j_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_himu_j_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_himu_j_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_himu_j_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_himu_j_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_himu_j_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_himu_j_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_himu_j_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_himu_j_340414

#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_lomu_a_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_lomu_a_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_lomu_a_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_lomu_a_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_lomu_a_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_lomu_a_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_lomu_a_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_lomu_a_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_lomu_a_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_lomu_a_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_lomu_a_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_lomu_a_340414

#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_lomu_b_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_lomu_b_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_lomu_b_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_lomu_b_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_lomu_b_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_lomu_b_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_lomu_b_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_lomu_b_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_lomu_b_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_lomu_b_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_lomu_b_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_lomu_b_340414

#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_lomu_c_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_lomu_c_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_lomu_c_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_lomu_c_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_lomu_c_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_lomu_c_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_lomu_c_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_lomu_c_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_lomu_c_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_lomu_c_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_lomu_c_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_lomu_c_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_lomu_d_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_lomu_d_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_lomu_d_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_lomu_d_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_lomu_d_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_lomu_d_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_lomu_d_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_lomu_d_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_lomu_d_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_lomu_d_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_lomu_d_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_lomu_d_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_lomu_e_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_lomu_e_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_lomu_e_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_lomu_e_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_lomu_e_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_lomu_e_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_lomu_e_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_lomu_e_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_lomu_e_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_lomu_e_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_lomu_e_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_lomu_e_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_lomu_f_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_lomu_f_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_lomu_f_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_lomu_f_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_lomu_f_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_lomu_f_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_lomu_f_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_lomu_f_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_lomu_f_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_lomu_f_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_lomu_f_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_lomu_f_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_lomu_g_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_lomu_g_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_lomu_g_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_lomu_g_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_lomu_g_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_lomu_g_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_lomu_g_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_lomu_g_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_lomu_g_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_lomu_g_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_lomu_g_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_lomu_g_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_lomu_h_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_lomu_h_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_lomu_h_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_lomu_h_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_lomu_h_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_lomu_h_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_lomu_h_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_lomu_h_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_lomu_h_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_lomu_h_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_lomu_h_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_lomu_h_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_lomu_i_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_lomu_i_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_lomu_i_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_lomu_i_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_lomu_i_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_lomu_i_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_lomu_i_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_lomu_i_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_lomu_i_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_lomu_i_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_lomu_i_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_lomu_i_340414
#ARTIFICIAL SELECTION

# #
# sim10p_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_lomu_j_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_lomu_j_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_lomu_j_340414

##
# sim10p_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_lomu_j_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_lomu_j_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_lomu_j_341404

##NATURAL SELECTION

##
# sim10p_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_1of2_nat_lomu_j_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_1of2_nat_lomu_j_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_1of2_nat_lomu_j_340414

##
# sim10p_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10p_mono_nat_lomu_j_340414_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10p_mono_nat_lomu_j_340414_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10p_mono_nat_lomu_j_340414

#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_himu_a_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_himu_a_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_himu_a_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_himu_a_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_himu_a_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_himu_a_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_himu_a_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_himu_a_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_himu_a_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_himu_a_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_himu_a_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_himu_a_140434

#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_himu_b_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_himu_b_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_himu_b_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_himu_b_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_himu_b_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_himu_b_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_himu_b_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_himu_b_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_himu_b_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_himu_b_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_himu_b_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_himu_b_140434

#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_himu_c_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_himu_c_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_himu_c_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_himu_c_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_himu_c_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_himu_c_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_himu_c_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_himu_c_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_himu_c_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_himu_c_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_himu_c_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_himu_c_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_himu_d_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_himu_d_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_himu_d_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_himu_d_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_himu_d_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_himu_d_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_himu_d_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_himu_d_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_himu_d_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_himu_d_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_himu_d_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_himu_d_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_himu_e_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_himu_e_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_himu_e_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_himu_e_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_himu_e_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_himu_e_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_himu_e_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_himu_e_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_himu_e_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_himu_e_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_himu_e_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_himu_e_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_himu_f_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_himu_f_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_himu_f_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_himu_f_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_himu_f_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_himu_f_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_himu_f_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_himu_f_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_himu_f_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_himu_f_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_himu_f_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_himu_f_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_himu_g_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_himu_g_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_himu_g_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_himu_g_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_himu_g_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_himu_g_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_himu_g_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_himu_g_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_himu_g_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_himu_g_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_himu_g_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_himu_g_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_himu_h_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_himu_h_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_himu_h_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_himu_h_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_himu_h_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_himu_h_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_himu_h_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_himu_h_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_himu_h_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_himu_h_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_himu_h_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_himu_h_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_himu_i_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_himu_i_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_himu_i_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_himu_i_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_himu_i_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_himu_i_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_himu_i_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_himu_i_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_himu_i_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_himu_i_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_himu_i_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_himu_i_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_himu_j_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_himu_j_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_himu_j_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_himu_j_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_himu_j_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_himu_j_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_himu_j_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_himu_j_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_himu_j_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.005 # Forward Mutation Rate
v = 0.0005# Backward Mutation Rate
m = 0.0005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_himu_j_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_himu_j_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_himu_j_140434

#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_lomu_a_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_lomu_a_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_lomu_a_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_lomu_a_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_lomu_a_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_lomu_a_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_lomu_a_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_lomu_a_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_lomu_a_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_lomu_a_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_lomu_a_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_lomu_a_140434

#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_lomu_b_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_lomu_b_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_lomu_b_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_lomu_b_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_lomu_b_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_lomu_b_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_lomu_b_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_lomu_b_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_lomu_b_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_lomu_b_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_lomu_b_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_lomu_b_140434

#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_lomu_c_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_lomu_c_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_lomu_c_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_lomu_c_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_lomu_c_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_lomu_c_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_lomu_c_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_lomu_c_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_lomu_c_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_lomu_c_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_lomu_c_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_lomu_c_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_lomu_d_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_lomu_d_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_lomu_d_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_lomu_d_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_lomu_d_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_lomu_d_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_lomu_d_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_lomu_d_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_lomu_d_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_lomu_d_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_lomu_d_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_lomu_d_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_lomu_e_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_lomu_e_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_lomu_e_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_lomu_e_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_lomu_e_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_lomu_e_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_lomu_e_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_lomu_e_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_lomu_e_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_lomu_e_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_lomu_e_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_lomu_e_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_lomu_f_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_lomu_f_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_lomu_f_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_lomu_f_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_lomu_f_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_lomu_f_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_lomu_f_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_lomu_f_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_lomu_f_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_lomu_f_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_lomu_f_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_lomu_f_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_lomu_g_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_lomu_g_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_lomu_g_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_lomu_g_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_lomu_g_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_lomu_g_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_lomu_g_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_lomu_g_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_lomu_g_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_lomu_g_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_lomu_g_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_lomu_g_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_lomu_h_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_lomu_h_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_lomu_h_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_lomu_h_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_lomu_h_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_lomu_h_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_lomu_h_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_lomu_h_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_lomu_h_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_lomu_h_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_lomu_h_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_lomu_h_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_lomu_i_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_lomu_i_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_lomu_i_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_lomu_i_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_lomu_i_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_lomu_i_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_lomu_i_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_lomu_i_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_lomu_i_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_lomu_i_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_lomu_i_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_lomu_i_140434
#ARTIFICIAL SELECTION

# #
# sim10q_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM   	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_lomu_j_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_lomu_j_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_lomu_j_140434

##
# sim10q_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:41:29 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_lomu_j_341404_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_lomu_j_341404_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_lomu_j_341404

##NATURAL SELECTION

##
# sim10q_mono_1of2_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_1of2_nat_lomu_j_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_1of2_nat_lomu_j_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
		sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_1of2_nat_lomu_j_140434

##
# sim10q_mono_nat.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# # 1 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by natural selection (bottleneck) in 1 sub pop before splitting
# If other selection events before that at locus 4, this occurs before the split also

# genepop, fstat format
# 
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: April 11, 2013 01:50:21 PM 	   
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 1   # Initial Number of Subpopns
d = 6 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.00005 # Forward Mutation Rate
v = 0.000005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 60   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10q_mono_nat_lomu_j_140434_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10q_mono_nat_lomu_j_140434_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MapSelector(loci=4, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j1, end = (j1+d1)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
	    sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10q_mono_nat_lomu_j_140434

#ARTIFICIAL SELECTION

# #
# sim10n_mono_1of2.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 Initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM  	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.0005 # Forward Mutation Rate
u = 0.00005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_a_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_a_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_a_000000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_a_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_a_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_a_040000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_a_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_a_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
      sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_a_001400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_a_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_a_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
      sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_a_000034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_a_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_a_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_a_041400

###################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_a_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_a_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_a_041434
##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_a_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_a_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_a_040034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_a_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_a_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_a_140400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 80    # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_a_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_a_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_a_340004

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_a_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_a_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_a_341404

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 300   # Number of Steps (Generations) before selection event 2
j3 = 200   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_a_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_a_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_a_043414

##
# sim10n_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.0005 # Forward Mutation Rate
u = 0.00005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_b_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_b_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_b_000000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_b_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_b_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_b_040000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_b_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_b_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
      sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_b_001400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_b_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_b_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
      sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_b_000034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_b_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_b_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_b_041400

###################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_b_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_b_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = 500)

		
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_b_041434
##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_b_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_b_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_b_040034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_b_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_b_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_b_140400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 80    # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_b_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_b_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_b_340004

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_b_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_b_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_b_341404

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 300   # Number of Steps (Generations) before selection event 2
j3 = 200   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_b_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_b_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_b_043414

##
# sim10n_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.0005 # Forward Mutation Rate
u = 0.00005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_c_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_c_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_c_000000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_c_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_c_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_c_040000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_c_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_c_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
      sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_c_001400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_c_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_c_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
      sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_c_000034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_c_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_c_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_c_041400

###################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_c_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_c_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_c_041434
##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_c_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_c_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_c_040034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_c_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_c_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_c_140400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 80    # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_c_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_c_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_c_340004

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_c_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_c_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_c_341404

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 300   # Number of Steps (Generations) before selection event 2
j3 = 200   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_c_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_c_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_c_043414

##
# sim10n_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.0005 # Forward Mutation Rate
u = 0.00005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_d_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_d_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_d_000000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_d_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_d_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_d_040000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_d_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_d_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
      sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_d_001400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_d_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_d_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
      sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_d_000034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_d_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_d_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_d_041400

###################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_d_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_d_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_d_041434
##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_d_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_d_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_d_040034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_d_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_d_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_d_140400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 80    # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_d_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_d_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_d_340004

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_d_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_d_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_d_341404

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 300   # Number of Steps (Generations) before selection event 2
j3 = 200   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_d_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_d_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_d_043414

##
# sim10n_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.0005 # Forward Mutation Rate
u = 0.00005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_e_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_e_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_e_000000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_e_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_e_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_e_040000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_e_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_e_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
      sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_e_001400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_e_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_e_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
      sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_e_000034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_e_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_e_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_e_041400

###################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_e_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_e_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_e_041434
##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_e_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_e_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_e_040034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_e_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_e_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_e_140400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 80    # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_e_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_e_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_e_340004

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_e_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_e_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_e_341404

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 300   # Number of Steps (Generations) before selection event 2
j3 = 200   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_e_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_e_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_e_043414

##
# sim10n_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.0005 # Forward Mutation Rate
u = 0.00005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_f_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_f_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_f_000000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_f_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_f_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_f_040000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_f_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_f_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
      sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_f_001400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_f_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_f_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
      sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_f_000034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_f_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_f_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_f_041400

###################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_f_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_f_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_f_041434
##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_f_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_f_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_f_040034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_f_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_f_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_f_140400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 80    # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_f_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_f_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_f_340004

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_f_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_f_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_f_341404

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 300   # Number of Steps (Generations) before selection event 2
j3 = 200   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_f_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_f_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_f_043414

##
# sim10n_mono.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	  
#


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
# Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.0005 # Forward Mutation Rate
u = 0.00005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_g_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_g_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_g_000000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_g_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_g_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_g_040000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_g_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_g_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
      sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_g_001400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_g_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_g_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
      sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_g_000034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_g_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_g_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_g_041400

###################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_g_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_g_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_g_041434
##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_g_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_g_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_g_040034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_g_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_g_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_g_140400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 80    # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_g_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_g_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_g_340004

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_g_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_g_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_g_341404

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 300   # Number of Steps (Generations) before selection event 2
j3 = 200   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_g_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_g_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_g_043414

##
# sim10n_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.0005 # Forward Mutation Rate
u = 0.00005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_h_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_h_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_h_000000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_h_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_h_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_h_040000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_h_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_h_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
      sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_h_001400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_h_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_h_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
      sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_h_000034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_h_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_h_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_h_041400

###################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_h_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_h_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_h_041434
##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_h_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_h_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_h_040034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_h_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_h_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_h_140400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 80    # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_h_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_h_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_h_340004

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_h_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_h_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_h_341404

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 300   # Number of Steps (Generations) before selection event 2
j3 = 200   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_h_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_h_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_h_043414

# #
# sim10n_mono.py
# 
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
# 
# #
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	  
# #


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
# Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.0005 # Forward Mutation Rate
u = 0.00005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_i_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_i_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_i_000000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_i_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_i_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_i_040000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_i_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_i_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
      sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_i_001400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_i_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_i_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
      sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_i_000034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_i_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_i_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_i_041400

###################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_i_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_i_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_i_041434
##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_i_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_i_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_i_040034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_i_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_i_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_i_140400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 80    # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_i_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_i_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_i_340004

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_i_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_i_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_i_341404

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 300   # Number of Steps (Generations) before selection event 2
j3 = 200   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_i_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_i_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_i_043414

##
# sim10n_mono.py
#
# Script to simulate a population with n subpops of i individuals, with population rebound, 
# natural selection and sampling.  Selection occuring at different periods.
# 2 initial subpopns, expanding to 6
# output as genotypes for Powermarker
# Selection at locus 4 by exclusion of minor allele (bottleneck) before splitting
# If other selection events before that at locus 4, this occurs before the split also
# 
# genepop, fstat format
#
##
# Author: Richard Stephens
# Created: October 29, 2012  08:56:33 AM
# Modified: February 25, 2013 12:24:31 PM	  
##


import simuOpt
import simuPOP as sim
import math
from simuPOP.utils import export, saveCSV, Exporter, migrIslandRates, importPopulation, viewVars
from simuPOP.sampling import drawRandomSample

       
## Settings ###################################################################################
n = 2   # Initial Number of Subpopns
d = 3 # Divisor for subpopn splitting
i = 100000 # Number of Indivs/Subpopn
l = 10    # Number of Loci per Chromosome
c = 5 # Number of Chromosomes
g = 10 # Number of Steps (Generations) before expansion
t = 2000 # Total Number of Steps (Generations)
u = 0.0005 # Forward Mutation Rate
u = 0.00005# Backward Mutation Rate
m = 0.00005  # Migration rate
e = 98 # Proportion of selfing
s1 = 0.008 # selection coefficient
q = 100000 # Maximum Population Size
r1 = 2.5e-4 #mean selection intensity as on chr 3H (based on a rate of 0.4Mb/cM) ie.(1/4/100)

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)

def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_j_000000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_j_000000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_j_000000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_j_040000_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_j_040000_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_j_040000

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_j_001400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_j_001400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
      sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_j_001400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_j_000034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_j_000034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
# 		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
# 		sim.DiscardIf(True, subPops=[
# 	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
      sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_j_000034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_j_041400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_j_041400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_j_041400

###################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_j_041434_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_j_041434_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_j_041434
##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_j_040034_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_j_040034_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_j_040034

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 300   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_j_140400_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_j_140400_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
#       sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_j_140400

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 200   # Number of Steps (Generations) before selection event 2
j3 = 80    # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_j_340004_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_j_340004_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
#       sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_j_340004

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 80   # Number of Steps (Generations) before selection event 2
j3 = 60   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_j_341404_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_j_341404_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_j_341404

##################################################################################################

z1 = 120   # Number of Steps (Generations) before population splitting
j1 = 100   # Number of Steps (Generations) before selection event 1
j2 = 300   # Number of Steps (Generations) before selection event 2
j3 = 200   # Number of Steps (Generations) before selection event 3

d1 = 50 # duration of selection event 1 (in generations)
d2 = 50 # duration of selection event 2 (in generations)
d3 = 50 # duration of selection event 3 (in generations)

##################################################################################################

pop = sim.Population(size=[i]*n, loci=[l]*c, infoFields=['migrate_to','fitness'])
pop.setVirtualSplitter(sim.ProductSplitter([
    sim.AffectionSplitter()
    ])
)
z=20
def demo(pop,gen):
	global q, g
	rate=2**15
	# if .... stop growing
	if gen >= g and all([x < q  for x in pop.subPopSizes()]):
		return ([(x+rate) for x in pop.subPopSizes(ancGen=-1)])
	elif gen >= g and all([x > q  for x in pop.subPopSizes()]):
		return ([q for x in pop.subPopSizes(ancGen=-1)])
	else:
		return ([x for x in pop.subPopSizes(ancGen=-1)])
		
def sampleAndExport(pop):
    sz = pop.subPopSizes()
    new_sz = [x//200 for x in sz]
    sample = drawRandomSample(pop, new_sz)
    export(sample, format='genepop', output='sim10n_mono_1of2_j_043414_lo_m_sample_%d.gen' % pop.dvars().gen, adjust = 0, gui = False)
    export(sample, format='fstat', output='sim10n_mono_1of2_j_043414_lo_m_sample_%d.dat' % pop.dvars().gen, gui = False)
    return True

simu = sim.Simulator(pop, rep=1)
simu.evolve(
    initOps=[
	    sim.InitGenotype(freq=(0.5, 0.5)),
	],
	preOps=[
		sim.SNPMutator(u = u, v = v),
		sim.Migrator(rate=migrIslandRates(m, n), end = z1-1),
		sim.SplitSubPops(subPops = [k for k in (list(range(n)))], proportions=[(1.0/d)]*d, randomize=True, at=z1),
	    sim.Migrator(rate=migrIslandRates(m, (n*d)), begin = z1+1),
		sim.MaPenetrance(loci=4, penetrance=[1, 1, 0.0]),
		sim.DiscardIf(True, subPops=[
	        (0, 'Affected')], at = j1),
        sim.MapSelector(loci=14, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j2, end = (j2+d2)),# Positive Selection for allele 1 at locus given
        sim.MapSelector(loci=34, fitness = {(0,0):1-s1, (0,1):1-s1, (1,1):1}, begin = j3, end = (j3+d3)),# Positive Selection for allele 1 at locus given
    ],
	
    #mating scheme structure for sexless mating with variable proportion of selfing, with recombination
    matingScheme=sim.HeteroMating([
        sim.SelfMating(ops=sim.Recombinator(intensity=r1), weight=e),
        sim.HomoMating(
            chooser=sim.CombinedParentsChooser(
                sim.RandomParentChooser(),
                sim.RandomParentChooser()),
            generator=sim.OffspringGenerator(
                ops= [
	                sim.MendelianGenoTransmitter(),
	                sim.Recombinator(intensity=r1)
                ],
            ),
            weight=100-e)
        ],
    subPopSize=demo
    ),
	postOps=[
		sim.PyOperator(func=sampleAndExport, at = [500,t])
    ],
	gen=(t+1),
)
print 'all done'
##sim10n_mono_1of2_j_043414




